File: packages/billing/cpp/src/billing_manager.cpp
#include "biztools/billing/billing_manager.hpp"
#include <random>
#include <sstream>
#include <iomanip>
#include <chrono>

namespace biztools {
namespace billing {

BillingManager::BillingManager(std::shared_ptr<PaymentProcessor> payment_processor)
    : payment_processor_(std::move(payment_processor)) {}

std::string BillingManager::create_product(
    const std::string& name,
    const std::string& description,
    const std::vector<Price>& prices
) {
    Product product;
    product.id = generate_id("prod");
    product.name = name;
    product.description = description;
    product.active = true;
    product.prices = prices;

    // Assign IDs to prices if not already set
    for (auto& price : product.prices) {
        if (price.id.empty()) {
            price.id = generate_id("price");
        }
    }

    products_[product.id] = product;
    return product.id;
}

void BillingManager::update_product(
    const std::string& product_id,
    const nlohmann::json& updates
) {
    auto it = products_.find(product_id);
    if (it == products_.end()) {
        throw std::runtime_error("Product not found");
    }

    auto& product = it->second;
    if (updates.contains("name")) {
        product.name = updates["name"];
    }
    if (updates.contains("description")) {
        product.description = updates["description"];
    }
    if (updates.contains("active")) {
        product.active = updates["active"];
    }
    if (updates.contains("prices")) {
        product.prices.clear();
        for (const auto& price : updates["prices"]) {
            product.prices.push_back(price.get<Price>());
        }
    }
}

std::string BillingManager::create_customer(
    const std::string& email,
    const std::string& name,
    const std::string& payment_method
) {
    Customer customer;
    customer.id = generate_id("cust");
    customer.email = email;
    customer.name = name;
    customer.payment_method = payment_method;
    customer.active = true;
    customer.created_at = std::chrono::system_clock::now();

    customers_[customer.id] = customer;
    return customer.id;
}

void BillingManager::update_customer(
    const std::string& customer_id,
    const nlohmann::json& updates
) {
    auto it = customers_.find(customer_id);
    if (it == customers_.end()) {
        throw std::runtime_error("Customer not found");
    }

    auto& customer = it->second;
    if (updates.contains("email")) {
        customer.email = updates["email"];
    }
    if (updates.contains("name")) {
        customer.name = updates["name"];
    }
    if (updates.contains("payment_method")) {
        customer.payment_method = updates["payment_method"];
    }
    if (updates.contains("active")) {
        customer.active = updates["active"];
    }
}

std::string BillingManager::create_subscription(
    const std::string& customer_id,
    const std::string& product_id,
    const std::string& price_id
) {
    // Validate customer and product
    if (!customers_.contains(customer_id)) {
        throw std::runtime_error("Customer not found");
    }
    if (!products_.contains(product_id)) {
        throw std::runtime_error("Product not found");
    }

    // Create subscription
    Subscription sub;
    sub.id = generate_id("sub");
    sub.customer_id = customer_id;
    sub.product_id = product_id;
    sub.price_id = price_id;
    sub.status = "active";
    sub.current_period_start = std::chrono::system_clock::now();
    
    // Set period end based on price interval
    const auto& product = products_[product_id];
    const auto* price = std::find_if(
        product.prices.begin(),
        product.prices.end(),
        [&price_id](const Price& p) { return p.id == price_id; }
    );
    
    if (price == product.prices.end()) {
        throw std::runtime_error("Price not found");
    }

    // Calculate period end
    auto period_end = sub.current_period_start;
    if (price->interval == "month") {
        period_end += std::chrono::hours(24 * 30 * price->interval_count);
    } else if (price->interval == "year") {
        period_end += std::chrono::hours(24 * 365 * price->interval_count);
    }
    
    sub.current_period_end = period_end;
    sub.auto_renew = true;

    // Store subscription
    subscriptions_[sub.id] = sub;

    // Generate initial invoice
    generate_invoice(sub.id);

    return sub.id;
}

void BillingManager::cancel_subscription(
    const std::string& subscription_id,
    bool immediate
) {
    auto it = subscriptions_.find(subscription_id);
    if (it == subscriptions_.end()) {
        throw std::runtime_error("Subscription not found");
    }

    auto& sub = it->second;
    if (immediate) {
        sub.status = "canceled";
        sub.canceled_at = std::chrono::system_clock::now();
    } else {
        sub.auto_renew = false;
    }
}

void BillingManager::process_monthly_billing() {
    auto now = std::chrono::system_clock::now();

    for (auto& [id, sub] : subscriptions_) {
        if (sub.status != "active" || !sub.auto_renew) {
            continue;
        }

        if (is_subscription_due(sub)) {
            generate_invoice(id);
        }
    }
}

void BillingManager::generate_invoice(const std::string& subscription_id) {
    auto it = subscriptions_.find(subscription_id);
    if (it == subscriptions_.end()) {
        throw std::runtime_error("Subscription not found");
    }

    const auto& sub = it->second;
    const auto& product = products_[sub.product_id];
    const auto* price = std::find_if(
        product.prices.begin(),
        product.prices.end(),
        [&sub](const Price& p) { return p.id == sub.price_id; }
    );

    if (price == product.prices.end()) {
        throw std::runtime_error("Price not found");
    }

    Invoice invoice;
    invoice.id = generate_id("inv");
    invoice.customer_id = sub.customer_id;
    invoice.subscription_id = sub.id;
    invoice.amount = price->amount;
    invoice.currency = price->currency;
    invoice.status = "open";
    invoice.created_at = std::chrono::system_clock::now();

    invoices_[invoice.id] = invoice;
}

bool BillingManager::process_invoice(const std::string& invoice_id) {
    auto it = invoices_.find(invoice_id);
    if (it == invoices_.end()) {
        throw std::runtime_error("Invoice not found");
    }

    auto& invoice = it->second;
    if (invoice.status != "open") {
        return false;
    }

    // Process payment
    bool success = payment_processor_->process_payment(
        invoice.customer_id,
        invoice.amount,
        invoice.currency
    );

    if (success) {
        invoice.status = "paid";
        invoice.paid_at = std::chrono::system_clock::now();

        // Update subscription period if payment successful
        auto sub_it = subscriptions_.find(invoice.subscription_id);
        if (sub_it != subscriptions_.end()) {
            auto& sub = sub_it->second;
            sub.current_period_start = sub.current_period_end;
            sub.current_period_end += (sub.current_period_end - sub.current_period_start);
        }
    } else {
        // Handle failed payment
        auto sub_it = subscriptions_.find(invoice.subscription_id);
        if (sub_it != subscriptions_.end()) {
            auto& sub = sub_it->second;
            sub.status = "past_due";
        }
    }

    return success;
}

std::vector<Invoice> BillingManager::get_customer_invoices(
    const std::string& customer_id
) {
    std::vector<Invoice> result;
    for (const auto& [id, invoice] : invoices_) {
        if (invoice.customer_id == customer_id) {
            result.push_back(invoice);
        }
    }
    return result;
}

std::optional<Subscription> BillingManager::get_subscription(
    const std::string& subscription_id
) {
    auto it = subscriptions_.find(subscription_id);
    if (it != subscriptions_.end()) {
        return it->second;
    }
    return std::nullopt;
}

std::vector<Product> BillingManager::get_available_products() {
    std::vector<Product> result;
    for (const auto& [id, product] : products_) {
        if (product.active) {
            result.push_back(product);
        }
    }
    return result;
}

// Private helper methods
void BillingManager::validate_subscription(const Subscription& sub) {
    if (!customers_.contains(sub.customer_id)) {
        throw std::runtime_error("Invalid customer ID");
    }
    if (!products_.contains(sub.product_id)) {
        throw std::runtime_error("Invalid product ID");
    }
}

std::string BillingManager::generate_id(const std::string& prefix) {
    static std::random_device rd;
    static std::mt19937 gen(rd());
    static std::uniform_int_distribution<> dis(0, 35);

    std::stringstream ss;
    ss << prefix << "_";
    
    const char charset[] = "0123456789abcdefghijklmnopqrstuvwxyz";
    for (int i = 0; i < 16; ++i) {
        ss << charset[dis(gen)];
    }
    
    return ss.str();
}

bool BillingManager::is_subscription_due(const Subscription& sub) {
    auto now = std::chrono::system_clock::now();
    return now >= sub.current_period_end;
}

} // namespace billing
} // namespace biztools 

File: packages/billing/biztools/billing/billing_manager.py
from datetime import datetime
from typing import List, Optional, Dict, Any
import logging
from .stripe_processor import StripeProcessor

logger = logging.getLogger(__name__)

class BillingManager:
    def __init__(self, payment_processor: StripeProcessor):
        self.payment_processor = payment_processor
        self.customers = {}
        self.products = {}
        self.subscriptions = {}
        self.invoices = {}

    def create_customer(self, email: str, name: str, payment_method: str) -> str:
        # Check if customer already exists
        for customer_id, customer in self.customers.items():
            if customer["email"] == email:
                logger.info(f"Customer already exists: {customer_id}")
                return customer_id

        # Create new customer
        customer_id = self.payment_processor.create_stripe_customer(email, payment_method)
        self.customers[customer_id] = {
            "id": customer_id,
            "email": email,
            "name": name,
            "payment_method": payment_method,
            "created_at": datetime.now()
        }
        logger.info(f"Created new customer: {customer_id}")
        return customer_id

    def create_product(self, name: str, description: str, prices: List[Dict[str, Any]]) -> str:
        product_id = f"prod_{len(self.products)}"
        self.products[product_id] = {
            "id": product_id,
            "name": name,
            "description": description,
            "prices": [{
                "id": f"price_{len(self.products)}_{i}",
                **price
            } for i, price in enumerate(prices)]
        }
        logger.info(f"Created product: {product_id}")
        return product_id

    def create_subscription(self, customer_id: str, product_id: str, price_id: str) -> str:
        if customer_id not in self.customers:
            raise ValueError("Customer not found")
        if product_id not in self.products:
            raise ValueError("Product not found")

        subscription_id = f"sub_{len(self.subscriptions)}"
        self.subscriptions[subscription_id] = {
            "id": subscription_id,
            "customer_id": customer_id,
            "product_id": product_id,
            "price_id": price_id,
            "status": "active",
            "created_at": datetime.now()
        }
        logger.info(f"Created subscription: {subscription_id}")

        # Create initial invoice
        self.generate_invoice(subscription_id)
        return subscription_id

    def generate_invoice(self, subscription_id: str) -> str:
        if subscription_id not in self.subscriptions:
            raise ValueError("Subscription not found")

        sub = self.subscriptions[subscription_id]
        product = self.products[sub["product_id"]]
        price = next(p for p in product["prices"] if p["id"] == sub["price_id"])

        invoice_id = f"inv_{len(self.invoices)}"
        self.invoices[invoice_id] = {
            "id": invoice_id,
            "customer_id": sub["customer_id"],
            "subscription_id": subscription_id,
            "amount": price["amount"],
            "currency": price["currency"],
            "status": "open",
            "created_at": datetime.now(),
            "paid_at": None
        }
        logger.info(f"Generated invoice: {invoice_id}")
        return invoice_id

    def process_invoice(self, invoice_id: str) -> bool:
        if invoice_id not in self.invoices:
            logger.error(f"Invoice not found: {invoice_id}")
            raise ValueError("Invoice not found")

        invoice = self.invoices[invoice_id]
        if invoice["status"] != "open":
            logger.info(f"Invoice {invoice_id} is not open (status: {invoice['status']})")
            return False

        success = self.payment_processor.process_payment(
            invoice["customer_id"],
            invoice["amount"],
            invoice["currency"]
        )

        if success:
            invoice["status"] = "paid"
            invoice["paid_at"] = datetime.now()
            logger.info(f"Invoice {invoice_id} processed successfully")
        else:
            logger.error(f"Failed to process invoice {invoice_id}")

        return success

    def get_customer_invoices(self, customer_id: str) -> List[Dict[str, Any]]:
        invoices = [
            invoice for invoice in self.invoices.values()
            if invoice["customer_id"] == customer_id
        ]
        logger.info(f"Found {len(invoices)} invoices for customer {customer_id}")
        return sorted(invoices, key=lambda x: x["created_at"], reverse=True)

    def get_subscription(self, subscription_id: str) -> Optional[Dict[str, Any]]:
        return self.subscriptions.get(subscription_id)

    def get_available_products(self) -> List[Dict[str, Any]]:
        return list(self.products.values())

    def delete_invoice(self, invoice_id: str) -> bool:
        if invoice_id not in self.invoices:
            logger.error(f"Invoice not found: {invoice_id}")
            return False

        invoice = self.invoices[invoice_id]
        if invoice["status"] == "paid":
            # Refund the payment
            logger.info(f"Refunding payment for invoice {invoice_id}")
            success = self.payment_processor.refund_payment(invoice_id)
            if not success:
                logger.error(f"Failed to refund payment for invoice {invoice_id}")
                return False

        # Delete the invoice
        del self.invoices[invoice_id]
        logger.info(f"Deleted invoice {invoice_id}")

        # Cancel subscription if it exists
        subscription_id = invoice.get("subscription_id")
        if subscription_id and subscription_id in self.subscriptions:
            self.subscriptions[subscription_id]["status"] = "canceled"
            logger.info(f"Canceled subscription {subscription_id}")

        return True 